docs/rfcs/rfc-035-authorization.md

```diff
@@ -0,0 +1,759 @@
+
+```gql
+input JWTPayloadWhere {
+    AND: [JWTPayloadWhere!]
+    NOT: JWTPayloadWhere
+    OR: [JWTPayloadWhere!]
+    sub: String
+    sub_CONTAINS: String
+    sub_ENDS_WITH: String
+    sub_IN: [String!]
+    sub_STARTS_WITH: String
+    roles: [String!]
+    roles_INCLUDES: String
+}
+```
+
+### Authorization directive definitions
+
+Firstly, for `User`:
+
+```gql
+input UserAuthorizationWhere {
+  OR: [UserAuthorizationWhere!]
+  AND: [UserAuthorizationWhere!]
+  NOT: UserAuthorizationWhere
+  jwtPayload: JWTPayloadWhere
+  node: UserWhere
+}
+
+enum AuthorizationFilterOperation {
+  READ
+  UPDATE
+  DELETE
+  CREATE_RELATIONSHIP
+  DELETE_RELATIONSHIP
+}
+
+input UserAuthorizationFilterRule {
+  operations: [AuthorizationFilterOperation!]! = [READ, UPDATE, DELETE, CREATE_RELATIONSHIP, DELETE_RELATIONSHIP]
+  requireAuthentication: Boolean! = true
+  where: UserAuthorizationWhere!
+}
+
+enum AuthorizationValidateBefore {
+  READ
+  CREATE
+  UPDATE
+  DELETE
+  CREATE_RELATIONSHIP
+  DELETE_RELATIONSHIP
+}
+
+enum AuthorizationValidateAfter {
+  CREATE
+  UPDATE
+  DELETE
+  CREATE_RELATIONSHIP
+  DELETE_RELATIONSHIP
+}
+
+input UserAuthorizationFilterRule {
+  before: [AuthorizationValidateBefore!]! = [READ, CREATE, UPDATE, DELETE, CREATE_RELATIONSHIP, DELETE_RELATIONSHIP]
+  after: [AuthorizationValidateAfter!]! = [CREATE, UPDATE, DELETE, CREATE_RELATIONSHIP, DELETE_RELATIONSHIP]
+  requireAuthentication: Boolean! = true
+  where: UserAuthorizationWhere!
+}
+
+directive @authorization(
+  filter: [UserAuthorizationFilterRule!]
+  validate: [UserAuthorizationValidateRule!]!
+) on OBJECT | FIELD_DEFINITION | INTERFACE
+```
+
+And for `Post`:
+
+```gql
+input PostAuthorizationWhere {
+  OR: [PostAuthorizationWhere!]
+  AND: [PostAuthorizationWhere!]
+  NOT: PostAuthorizationWhere
+  jwtPayload: JWTPayloadWhere
+  node: PostWhere
+}
+
+enum AuthorizationFilterOperation {
+  READ
+  UPDATE
+  DELETE
+  CREATE_RELATIONSHIP
+  DELETE_RELATIONSHIP
+}
+
+input PostAuthorizationFilterRule {
+  operations: [AuthorizationFilterOperation!]! = [READ, UPDATE, DELETE, CREATE_RELATIONSHIP, DELETE_RELATIONSHIP]
+  requireAuthentication: Boolean! = true
+  where: PostAuthorizationWhere!
+}
+
+enum AuthorizationValidateBefore {
+  READ
+  CREATE
+  UPDATE
+  DELETE
+  CREATE_RELATIONSHIP
+  DELETE_RELATIONSHIP
+}
+
+enum AuthorizationValidateAfter {
+  CREATE
+  UPDATE
+  DELETE
+  CREATE_RELATIONSHIP
+  DELETE_RELATIONSHIP
+}
+
+input PostAuthorizationFilterRule {
+  before: [AuthorizationValidateBefore!]! = [READ, CREATE, UPDATE, DELETE, CREATE_RELATIONSHIP, DELETE_RELATIONSHIP]
+  after: [AuthorizationValidateAfter!]! = [CREATE, UPDATE, DELETE, CREATE_RELATIONSHIP, DELETE_RELATIONSHIP]
+  requireAuthentication: Boolean! = true
+  where: PostAuthorizationWhere!
+}
+
+directive @authorization(
+  filter: [PostAuthorizationFilterRule!]
+  validate: [PostAuthorizationValidateRule!]!
+) on OBJECT | FIELD_DEFINITION | INTERFACE
```

Comment:
Is there anyway we could provide this or a similar directive for `SCHEMA` level rules? Obviously, we'd only be able to check the jwt at that level but I know this is something customers have asked for

Reply:
Even if this is out of scope here, it's probably worth noting this in an out of scope section

Reply:
My intention is to explore GraphQL Shield for this kind of thing... 👀